### ISSUE TITLE
apt_key --check falsely reports change

### ISSUE THREAD

**Issue by t2d** [ansible/ansible-modules-core#2622]
Issue Type: Bug Report
Ansible: Version: ansible 1.9.4
Environment: Debian Wheezy
Summary:
Task:

```
- name: download and import tor signing key
  apt_key:
    keyserver: keys.gnupg.net
    id: A3C4F0F979CAA22CDBA8F512EE8CBC9E886DDD89
    state: present
```

If I call it normally, it returns "ok".
If I run it with --check, it always falsely returns "changed".

Expected Results:
It should return "ok" in both cases.

**Comment by canha** [comment]
I've experienced this behaviour as well - on both Wheezy and Jessie.

**Comment by praseodym** [comment]
This is because by default the apt_key module uses short key ids:
https://github.com/ansible/ansible-modules-core/blob/35b131e71a06ac52271782b03151c35ab4a7e219/packaging/os/apt_key.py#L238

**Comment by gmangin** [comment]
Issue Type: Bug Report
Ansible: Version: ansible 2.0.1
Environment: Debian jessie
Summary:
Task:

```
- name: download and import key without id
  apt_key:
    url=http://debian.datastax.com/debian/repo_key
    state=present
```

If I call it normally, it returns "ok".
If I run it with --check, it always falsely returns "changed".

Expected Results:
It should return "ok" in both cases.

I saw the problem in the code -> https://github.com/ansible/ansible-modules-core/blob/35b131e71a06ac52271782b03151c35ab4a7e219/packaging/os/apt_key.py#L251

**Comment by selivan** [comment]
Here is another issue with false positive `--check` for apt_key: https://github.com/ansible/ansible-modules-core/issues/3958 but there `data=` is used. I suppose, this issues are realted.

**Comment by nitzmahone** [comment]
This is arguably a feature request, as the apt_key module doesn't currently have a mechanism to parse the keys (it'll do what you expect if you include the key_id). Someone would need to add GPG key parsing to the module to implement the desired behavior.

**Comment by t2d** [comment]
@nitzmahone no, it is not a feature request. I included the id, but the module doesn't recognize the long id. Short key ids are unsafe and should not be used.

**Comment by ansibot** [comment]
@jvantuyl, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by jvantuyl** [comment]
@t2d We defer to apt-key to actually handle the keys.  We would like to forbid short keys and use long keys only.  Unfortunately, we're blocked on this bug:  https://bugs.launchpad.net/ubuntu/+source/apt/+bug/1481871

When that's fixed, we'll drop short key support entirely.  This will probably break everyone, but it's security-critical so I'm fine with that.

**Comment by t2d** [comment]
@jvantuyl yes, that sounds very reasonable. short key collision is a real thing.

**Comment by ansibot** [comment]
@jvantuyl, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@jvantuyl, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jvantuyl** [comment]
needs_contributor

**Comment by abadger** [comment]
Pushed a fix for this to https://github.com/ansible/ansible-modules-core/pull/5353  it is a little unrelated but it builds on the existing cleanups there so it makes sense to combine it and backport it together.
